src/js/autocomplete.js

```diff
@@ -1,14 +1,247 @@
+		}
+		return [character, shoudHighlight];
+	});
+}
+
+/**
+ * Create the DOM tree which corresponds to
+ * <div class="o-autocomplete__menu-loading-container">
+ * 	<div class="o-autocomplete__menu-loading"></div>
+ * </div>
+ * @returns {HTMLDivElement} The loading container as a HTMLDivElement
+ */
+function createLoadingContainer() {
+	const loadingContainer = document.createElement('div');
+	loadingContainer.classList.add('o-autocomplete__menu-loading-container');
+	const loading = document.createElement('div');
+	loading.classList.add('o-autocomplete__menu-loading');
+	loadingContainer.appendChild(loading);
+	return loadingContainer;
+}
+
+/**
+ * @callback PopulateResults
+ * @param {Array<string>} results - The results to show in the suggestions dropdown
+ * @returns {void}
+ */
+
+/**
+ * @callback Suggestions
+ * @param {string} query - Text which was typed into the autocomplete by the user
+ * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+ * @returns {void}
+*/
+
+
+/**
+ * @typedef {Object} AutocompleteOptions
+ * @property {Suggestions} source - The function which retrieves the suggestions to display
+ */
+
 class Autocomplete {
 	/**
 	 * Class constructor.
 	 * @param {HTMLElement} [autocompleteEl] - The component element in the DOM
-	 * @param {Object} [options={}] - An options object for configuring the component
+	 * @param {AutocompleteOptions} [options={}] - An options object for configuring the component
 	 */
 	constructor (autocompleteEl, options) {
 		this.autocompleteEl = autocompleteEl;
-		this.options = Object.assign({}, {
+		this.options = Object.assign({
+			placeholder: '',
+			cssNamespace: 'o-autocomplete',
+			displayMenu: 'overlay',
+			showNoOptionsFound: false,
+			templates: {
+				suggestion: this.suggestionTemplate.bind(this)
+			}
 		}, options || Autocomplete.getDataAttributes(autocompleteEl));
+
+		if (this.options.source) {
+			// If source is a string, then it is the name of a global function to use.
+			// If source is not a string, then it is a function to use.
+			/**
+			 * @type {Suggestions}
+			 */
+			const customSuggestions = typeof this.options.source === 'string' ? window[this.options.source] : this.options.source;
+			/**
+			 * @param {string} query - Text which was typed into the autocomplete by the user
+			 * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+			 * @returns {void}
+			*/
+			this.options.source = (query, populateResults) => {
+				this.showLoadingPane();
+				/**
+				 * @param {Array<string>} results - The results to show in the suggestions dropdown
+				 * @returns {void}
+				 */
+				const callback = (results) => {
+					this.hideLoadingPane();
+					populateResults(results);
+				};
+				customSuggestions(query, callback);
+			};
+			const input = autocompleteEl.querySelector('input');
+			const id = input.getAttribute('id');
```

Comment:
Question: is this id required for accessible-autocomplete to work? If so should we throw our own error here?

Reply:
It is required, accessible-autocomplete throws an error if it is missing. We can throw our own error instead, yeah 👍

Reply:
Done in [`da94664` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/da9466432c1b585cde53506a6488f8fdd228bfea)